network/retrieve/retrieve.go

```diff
@@ -0,0 +1,432 @@
+	defer cancel()
+
+	req := &storage.Request{
+		Addr:   msg.Addr,
+		Origin: p.ID(),
+	}
+	chunk, err := r.netStore.Get(ctx, chunk.ModeGetRequest, req)
+	if err != nil {
+		retrieveChunkFail.Inc(1)
+		p.logDebug("netstore.Get can not retrieve chunk", "ref", msg.Addr, "err", err)
+		return
+	}
+
+	p.logTrace("retrieval.handleRetrieveRequest - delivery", "ref", msg.Addr)
+
+	deliveryMsg := &ChunkDelivery{
+		Addr:  chunk.Address(),
+		SData: chunk.Data(),
+	}
+
+	err = p.Send(ctx, deliveryMsg)
+	if err != nil {
+		p.logError("retrieval.handleRetrieveRequest - peer delivery failed", "ref", msg.Addr, "err", err)
+		osp.LogFields(olog.Bool("delivered", false))
+		return
+	}
+	osp.LogFields(olog.Bool("delivered", true))
+}
+
+// handleChunkDelivery handles a ChunkDelivery message from a certain peer
+// if the chunk proximity order in relation to our base address is within depth
+// we treat the chunk as a chunk received in syncing
+func (r *Retrieval) handleChunkDelivery(ctx context.Context, p *Peer, msg *ChunkDelivery) error {
+	p.logDebug("retrieval.handleChunkDelivery", "ref", msg.Addr)
+	var osp opentracing.Span
+	ctx, osp = spancontext.StartSpan(
+		ctx,
+		"handle.chunk.delivery")
+
+	processReceivedChunksCount.Inc(1)
+
+	// record the last time we received a chunk delivery message
+	lastReceivedRetrieveChunksMsg.Update(time.Now().UnixNano())
+
+	// count how many chunks we receive for retrieve requests per peer
+	peermetric := fmt.Sprintf("chunk.delivery.%x", p.BzzAddr.Over()[:16])
+	metrics.GetOrRegisterCounter(peermetric, nil).Inc(1)
+
+	peerPO := chunk.Proximity(p.BzzAddr.Over(), msg.Addr)
+	po := chunk.Proximity(r.kad.BaseAddr(), msg.Addr)
+	depth := r.kad.NeighbourhoodDepth()
+	var mode chunk.ModePut
+	// chunks within the area of responsibility should always sync
+	// https://github.com/ethersphere/go-ethereum/pull/1282#discussion_r269406125
+	if po >= depth || peerPO < po {
+		mode = chunk.ModePutSync
+	} else {
+		// do not sync if peer that is sending us a chunk is closer to the chunk then we are
+		mode = chunk.ModePutRequest
+	}
+
+	p.logTrace("handle.chunk.delivery", "ref", msg.Addr)
+
+	go func() {
+		defer osp.Finish()
+		p.logTrace("handle.chunk.delivery", "put", msg.Addr)
+		_, err := r.netStore.Put(ctx, mode, storage.NewChunk(msg.Addr, msg.SData))
+		if err != nil {
+			if err == storage.ErrChunkInvalid {
+				p.Drop()
+			}
+		}
+		p.logTrace("handle.chunk.delivery", "done put", msg.Addr, "err", err)
+	}()
+	return nil
+}
+
+// RequestFromPeers sends a chunk retrieve request to the next found peer
+func (r *Retrieval) RequestFromPeers(ctx context.Context, req *storage.Request, localID enode.ID) (*enode.ID, error) {
+	log.Debug("retrieval.requestFromPeers", "req.Addr", req.Addr)
+	metrics.GetOrRegisterCounter("network.retrieve.requestfrompeers", nil).Inc(1)
+
+	sp, err := r.findPeer(ctx, req)
+	if err != nil {
+		log.Trace(err.Error())
+		return nil, err
+	}
+
+	protoPeer := r.getPeer(sp.ID())
```

Comment:
As @nonsense said in comment above, protoPeer can be nil. Maybe just to explicitly check for that before calling using it?

Reply:
ok i add a nil check. now the next question is how the nil should be handled. ideally we should not fail the request but maybe `goto` the line where `findPeer` is called

Reply:
I think you should fail the request if `getPeer` is nil, because the connection is already dropped due to another request/message problem.

Reply:
if a node requests a chunk from us, why should we fail because another node disconnected?

Reply:
Oh, that's right. Then I guess it is fine as you have written it.